File: C/2/4.c
#include <stdio.h>
#include <ctype.h>


/**
 * A function to find a character inside a string.
 *
 * @param char[] s  The null-terminated string to be searched
 * @param int    c  The character to be searched for
 *
 * @return int      A "boolean" indicating if the character was found or not.
 */
int find (char s[], int c)
{
    int i;

    for (i = 0; s[i] != '\0'; i++) {
        if (s[i] == c)
            return 1;
    }

    return 0;
}

/**
 * A function that deletes each character in s1 that matches any character in
 * the strings2.
 *
 * @param char[] s1 String to be squeezed
 * @param char[] s2 String to be checked against
 *
 * @return void
 */

void squeeze(char s1[], char s2[])
{
    int i, j;

    for (i = j = 0; s1[i] != '\0'; i++)
        if (!find(s2, s1[i]))
            s1[j++] = s1[i];
    s1[j] = '\0';
}

/**
 * A program to test string squeezing.
 */
main ()
{
    char first[] = "A string with multiple characters.";
    char second[] = "aeiou";

    printf("*** BEFORE ***\n");

    printf("\t'%s'\n\t'%s'\n", first, second);

    printf("\n*** AFTER ***\n");

    squeeze(first, second);

    printf("\t'%s'\n\t'%s'\n", first, second);
}

File: rustlang/2/4.rs
/// A function to find a character inside a string.

/// @param &str    s  A reference to the string to be searched
/// @param char    c  The character to be searched for

/// @return bool      A boolean indicating if the character was found or not.
fn find (s: &str, c: char) -> bool
{
    for _c in s.chars() {
        if _c == c {
            return true;
        }
    }

    return false;
}

/// A function that deletes each character in s1 that matches any character in
/// the strings2.

/// @param &String s1  A reference to the String to be squeezed
/// @param &str    s2  A reference to an immutable string slice with chars to be searched

/// @return String     The string resultant of all searches
fn squeeze(s1: &String, s2: &str) -> String
{
    // Because of Rust's way of dealing with strings, instead of modifying the
    // original one it is better to create a empty one and append all desired
    // characters. Rust strings are not the same thing as C array of chars.
    let mut result = String::with_capacity(s1.capacity());

    for _c in s1.chars() {
        // We don't really need this 'find' function, just kept it for no reason.
        if !find(s2, _c) {
            result.push(_c,);
        }
    }

    result
}

/// A program to test string squeezing.
fn main ()
{
    let first = String::from("A string with multiple characters.");
    let second = String::from("aeiou");

    print!("*** BEFORE ***\n");

    print!("\t'{}'\n\t'{}'\n", first, second);

    print!("\n*** AFTER ***\n");

    print!("\t'{}'\n\t'{}'\n", squeeze(&first, &second), second);
}

